src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs

```diff
@@ -3,140 +3,591 @@
+            // It must happen before we try to get the value task, since the task source is versioned and each instance must be awaited.
+            cancellationToken.ThrowIfCancellationRequested();
+        }
 
-        public override bool CanRead => _provider.CanRead;
+        // Concurrent call, this one lost the race.
+        if (!_sendTcs.TryGetValueTask(out ValueTask valueTask, this, cancellationToken))
+        {
+            throw new InvalidOperationException(SR.Format(SR.net_io_invalidnestedcall, "write"));
+        }
 
-        public bool ReadsCompleted => _provider.ReadsCompleted;
+        // No need to call anything since we already have a result, most likely an exception.
+        if (valueTask.IsCompleted)
+        {
+            return valueTask;
+        }
 
-        public override int Read(Span<byte> buffer) => _provider.Read(buffer);
+        // For an empty buffer complete immediately, close the writing side of the stream if necessary.
+        if (buffer.IsEmpty)
+        {
+            _sendTcs.TrySetResult();
+            if (completeWrites)
+            {
+                CompleteWrites();
+            }
+            return valueTask;
+        }
 
-        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default) => _provider.ReadAsync(buffer, cancellationToken);
+        try
+        {
+            _sendBuffers.Initialize(buffer);
+            unsafe
+            {
+                ThrowHelper.ThrowIfMsQuicError(MsQuicApi.Api.ApiTable->StreamSend(
+                    _handle.QuicHandle,
+                    _sendBuffers.Buffers,
+                    (uint)_sendBuffers.Count,
+                    completeWrites ? QUIC_SEND_FLAGS.FIN : QUIC_SEND_FLAGS.NONE,
+                    null));
+            }
+        }
+        catch (Exception ex)
+        {
+            _sendTcs.TrySetException(ex, final: true);
+            _sendBuffers.Reset();
+            throw;
+        }
 
-        public override bool CanWrite => _provider.CanWrite;
+        return valueTask;
+    }
 
-        public override void Write(ReadOnlySpan<byte> buffer) => _provider.Write(buffer);
 
-        public override bool CanTimeout => _provider.CanTimeout;
+    public void Abort(QuicAbortDirection abortDirection, long errorCode)
+    {
+        if (_disposed == 1)
+        {
+            return;
+        }
 
-        public override int ReadTimeout
+        QUIC_STREAM_SHUTDOWN_FLAGS flags = QUIC_STREAM_SHUTDOWN_FLAGS.NONE;
+        if (abortDirection.HasFlag(QuicAbortDirection.Read))
         {
-            get => _provider.ReadTimeout;
-            set => _provider.ReadTimeout = value;
+            flags |= QUIC_STREAM_SHUTDOWN_FLAGS.ABORT_RECEIVE;
+            if (_receiveTcs.TrySetException(ThrowHelper.GetOperationAbortedException("Read was aborted"), final: true))
```

Comment:
I added string resources for theses.
```suggestion
            if (_receiveTcs.TrySetException(ThrowHelper.GetOperationAbortedException(SR.net_quic_reading_aborted), final: true))
```

Reply:
Bad merge, fixing.